README
Apache::Singleton is Copyright (C) 2009, Michael Schout.

WHAT IS THIS?

Apache::Singleton works the same as Class::Singleton, but with various object
lifetime (scope) under mod_perl or mod_perl2. See the Class::Singleton manpage
first.

INSTALLATION

To install this module type the following:

    perl Makefile.PL
    make
    make test
    make install

DEPENDENCIES

This module requires these other CPAN modules:

    Test::More 0.32 or later
    mod_perl or mod_perl2

COPYRIGHT AND LICENCE

   Copyright (C) 2009 Michael Schout.

   This program is free software; you can redistribute it and/or modify it
   under the terms of either:

   *   the GNU General Public License as published by the Free Software
       Foundation; either version 1, or (at your option) any later
       version, or

   *   the Artistic License version 2.0.


Changes
Revision history for Perl extension Apache::Singleton.

{{$NEXT}}

0.17  2017-08-02
    - Kwalitee: include META.json, declare minimum perl version, enable warnings

0.16  2014-11-07
    - fix signature test due to MYMETA.* in new EU::MM
    - Work around global destruction order issue by integrating on proposed
      patch to Class::Singleton [#68534] (Thanks Steve Hay!)

0.15  2012-04-02
    - fixed test: limit MaxClient to 1. process/printer_instance.t expects the
      same apache to handle each request. [RT 76244]

0.14  2011-12-06
    - load Apache::Singleton::Request if necessary [RT 71218]
    - note that PerlOptions +GlobalRequest is required for
      Apache::Singleton::Request

0.13  2011-05-29
    - No changes. Rebuild tarball in WinZip compatible way.

0.12  2011-05-28
    - Use Dist::Zilla for distribution building
    - Copy args passed to instance() to constructor

0.11  Tue Nov 24 2009
    - added real test suite using Apache::Test
    - reorganized test files
    - removed _delegator() from Apache::Singleton.  it causes problems under
      MP2

0.10  Sun Nov 22 2009
  *** new maintainer: Michael Schout <mschout@cpan.org>
    - updated Makefile.PL
    - added META_MERGE to Makefile.PL
    - update so mod_perl2 2.0 >= RC5 is supported
    - split out test mocks into t/lib
    - add mod_perl2 tests
    - sign the dist, add signature test
    - check for mod_perl version in Makefile.PL.
        if mod_perl2, require >= 1.9922
    - update README
    - remove META.yml from repo, make dist generates it

0.06  Tue Oct 19 17:29:05 JST 2004
    - removed dependency on Cache::Cache in Makefile.PL
    * Added mod_perl2 support
      Thanks to Frank Wiles <frank@wiles.org>

0.05  Tue Dec 25 05:59:55 JST 2001
    - Compatibility: defaults to Process in non-mod_perl environment

0.04  Sun Dec 23 21:20:04 JST 2001
    - Fixed docs
    - pulled off Server subclass: it was completely broken

0.03  Sat Dec 22 22:29:50 JST 2001
    - Added test for multiple classes
    * Added Request, Process, Server subclasses

0.02  Sat Dec 22 16:58:34 JST 2001
    - use pnotes instead of register_cleanup
      (Thanks to Perrin Harkins <perrin@elem.com>)

0.01  Thu Dec 20 12:45:29 2001
    - original version; created by h2xs 1.19



lib/Apache/Singleton.pm
# COPYRIGHT

package Apache::Singleton;

# ABSTRACT: Singleton class for mod_perl

use strict;
use warnings;

# load appropriate subclass
if ($ENV{MOD_PERL}) {
    require Apache::Singleton::Request;
}
else {
    require Apache::Singleton::Process;
}

sub instance {
    my $class = shift;

    my $instance = $class->_get_instance;
    unless (defined $instance) {
        $instance = $class->_new_instance(@_);
        $class->_set_instance($instance);
    }
    return $instance;
}

sub _new_instance {
    my $class = shift;

    my %args = (@_ && ref $_[0] eq 'HASH') ? %{ $_[0] } : @_;

    bless { %args }, $class;
}

# Abstract methods, but compatible default
sub _get_instance {
    my $class = shift;

    if ($ENV{MOD_PERL}) {
        $class->Apache::Singleton::Request::_get_instance(@_);
    }
    else {
        $class->Apache::Singleton::Process::_get_instance(@_);
    }
}

sub _set_instance {
    my $class = shift;

    if ($ENV{MOD_PERL}) {
        $class->Apache::Singleton::Request::_set_instance(@_);
    }
    else {
        $class->Apache::Singleton::Process::_set_instance(@_);
    }
}

1;
__END__

=head1 SYNOPSIS

  package Printer;
  # default:
  #   Request for mod_perl env
  #   Process for non-mod_perl env
  use base qw(Apache::Singleton);

  package Printer::PerRequest;
  use base qw(Apache::Singleton::Request);

  package Printer::PerProcess;
  use base qw(Apache::Singleton::Process);

=head1 DESCRIPTION

Apache::Singleton works the same as Class::Singleton, but with
various object lifetime (B<scope>). See L<Class::Singleton> first.

=head1 OBJECT LIFETIME

By inheriting one of the following sublasses of Apache::Singleton,
you can change the scope of your object.

=over 4

=item Request

  use base qw(Apache::Singleton::Request);

One instance for one request. Apache::Singleton will remove instance
on each request. Implemented using mod_perl C<pnotes> API. In mod_perl
environment (where C<$ENV{MOD_PERL}> is defined), this is the default
scope, so inheriting from Apache::Singleton would do the same effect.

B<NOTE>: You need C<PerlOptions +GlobalRequest> in your apache
configuration in order to use the I<Request> lifetime method.

=item Process

  use base qw(Apache::Singleton::Process);

One instance for one httpd process. Implemented using package
global. In non-mod_perl environment, this is the default scope, and
you may notice this is the same beaviour with Class::Singleton ;)

So you can use this module safely under non-mod_perl environment.

=back

=head1 CREDITS

Original idea by Matt Sergeant E<lt>matt@sergeant.orgE<gt> and Perrin
Harkins E<lt>perrin@elem.comE<gt>.

Initial implementation and versions 0.01 to 0.07 by Tatsuhiko Miyagawa
E<lt>miyagawa@bulknews.netE<gt>.

=head1 SEE ALSO

L<Apache::Singleton::Request>, L<Apache::Singleton::Process>,
L<Class::Singleton>

=cut


lib/Apache/Singleton/Request.pm
# COPYRIGHT

package Apache::Singleton::Request;

# ABSTRACT: One instance per One Request

use strict;
use warnings;

use base 'Apache::Singleton';

BEGIN {
    use constant MP2 => $mod_perl::VERSION >= 1.99 ? 1 : 0;

    if (MP2) {
        require Apache2::RequestUtil;
    }
    else {
        require Apache;
    }
}

sub _get_instance {
    my $class = shift;
    my $r = MP2 ? Apache2::RequestUtil->request : Apache->request;
    my $key = "apache_singleton_$class";
    return $r->pnotes($key);
}

sub _set_instance {
    my($class, $instance) = @_;
    my $r = MP2 ? Apache2::RequestUtil->request : Apache->request;
    my $key = "apache_singleton_$class";
    $r->pnotes($key => $instance);
}

1;

__END__

=head1 SYNOPSIS

  # in httpd.conf
  PerlOptions +GlobalRequest

  # in your module (e.g.: Printer.pm)
  package Printer;
  use base qw(Apache::Singleton::Request);

=head1 DESCRIPTION

See L<Apache::Singleton>.

=head1 SEE ALSO

L<Apache::Singleton>

=cut


Makefile.PL
use ExtUtils::MakeMaker;

my %conf = (
    NAME         => 'Apache::Singleton',
    VERSION_FROM => 'lib/Apache/Singleton.pm',
    PREREQ_PM    => {
        'Test::More' => 0.32,
        'mod_perl2'   => 0,
    }
);

if (mod_perl_version() == 2) {
    # require 2.0 RC5 (Apache -> Apache2)
    $conf{PREREQ_PM}{mod_perl2} = '1.9922';
}
else {
    $conf{PREREQ_PM}{mod_perl} = 0;
}

if (MM->can('signature_target')) {
    $conf{SIGN} = 1;
}

if ($] >= 5.005) {
    $conf{ABSTRACT_FROM} = 'lib/Apache/Singleton.pm';
    $conf{AUTHOR}        = 'Michael Schout <mschout@cpan.org>';
}

if ($ExtUtils::MakeMaker::VERSION >= 6.48) {
    $conf{META_MERGE} = {
        no_index => { directory => [qw(t)] },
        resources => {
            bugtracker => 'http://rt.cpan.org/NoAuth/Bugs.html?Dist=Apache-Singleton',
            repository => 'http://github.com/mschout/apache-singleton',
            homepage   => 'http://github.com/mschout/apache-singleton/downloads'
        }
    };
}

WriteMakefile(%conf);

sub mod_perl_version {
    # try to figure out what version of mod_perl is installed.
    eval {
        require mod_perl
    };
    unless ($@) {
        if ($mod_perl::VERSION >= 1.99) {
            # this is mod_perl2 < 2.0 RC5 (1.99_21 or earlier).
            # Apache renamed to Apache2 in 1.9922
            die "mod_perl 2.0.0 RC5 (1.9922) or later is required for this module (found $mod_perl::VERSION)";
        }

        return 1;
    }

    eval {
        require mod_perl2;
    };
    unless ($@) {
        return 2;
    }

    # we didn't fine a supported version issue a warning, and assume version 2.
    warn "no supported mod_perl version was found.  assuming version 2\n";

    return 2;
}
